File: this-is-coding-test/5.DFS_BFS/4.미로_탈출.py
from collections import deque

maze = "111111 000001 111111 100000 111111"
directions = [[-1, 0], [0, -1], [1, 0], [0, 1]]


def to_matrix(str, row):
    data = [[]] * row

    for index, value in enumerate(str.split(' ')):
        data[index] = list(map(int, value))

    return data


def bfs(value):
    row, column = len(value.split(' ')), len(value.split(' ')[0])
    data = to_matrix(value, row)
    costs = [[0] * column for _ in range(row)]

    queue = deque([[0, 0]])
    cost = 1
    costs[0][0] = cost

    while queue:
        cost += 1
        [cx, cy] = queue.popleft()

        for [dx, dy] in directions:
            [nx, ny] = [cx+dx, cy+dy]
            if 0 <= nx < row and 0 <= ny < column:
                
                queue.append([nx, ny])

            print(queue)

        break


print(bfs(maze))

File: this-is-coding-test/5.DFS_BFS/4.미로_탈출.js
// maze = "111111 000001 111111 100000 111111"
// maze = "101010 111111 000001 111111 111111"
const maze = "11000111 01111101 00000011 11111010 10001110 11100000 11111111"
const result = 32;
const graph = maze.split(' ').map(item => item.split('').map(value => parseInt(value, 10)));
const next = [[1, 0], [0, 1], [-1, 0], [0, -1]];

const bfs = () => {

}

const handler = (graph) => {
  const queue = [];
  queue.push([0, 0])

  while (queue.length > 0) {
    const [cx, cy] = queue.shift();

    next.forEach(([x, y]) => {
      const [nx, ny] = [cx+x, cy+y]

      if (nx >= 0 && ny >= 0 && nx < graph.length && ny < graph[0].length && graph[nx][ny] === 1) {

        graph[nx][ny] = graph[cx][cy] + 1;
        queue.push([nx, ny])
      }
    })
  }

  return graph
}

handler(graph)
console.log(graph[graph.length-1][graph[0].length-1])



